### PR TITLE
Feature/bca/spaces dnd

### LINKED ISSUE #3501: User defined top level spaces ordering
as per https://github.com/matrix-org/matrix-doc/pull/3230

Option to re-order top level spaces with drag n drop

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #3501
Add support for top level space custom ordering

dnd

### CODE REVIEW

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                orderInfo.first to (orderCommand.find { it.spaceId == orderInfo.first }?.order ?: orderInfo.second)
+            }
+                    .sortedWith(TestSpaceComparator())
+
+    private fun List<Pair<String, String?>>.assertSpaceOrdered() : List<Pair<String, String?>> {
+        assertEquals(this, this.sortedWith(TestSpaceComparator()))
+        return this
+    }
+
+    class TestSpaceComparator : Comparator<Pair<String, String?>> {
```
Comment (Dominaezzz) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt:249]:
```kotlin
val TestSpaceComparator = compareBy<Pair<String, String?>, String?>(nullsLast()) { it.second }
		.thenBy { it.first }
```

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/space/SpaceOrderUtils.kt [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import org.matrix.android.sdk.api.util.StringOrderUtils
+
+object SpaceOrderUtils {
+
+    data class SpaceReOrderCommand(
+            val spaceId: String,
+            val order: String
+    )
+
+    fun orderCommandsForMove(orderedSpacesToOrderMap: List<Pair<String, String?>>, movedSpaceId: String, delta: Int): List<SpaceReOrderCommand> {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/space/SpaceOrderUtils.kt:28]: A small KDoc would be useful here :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/StringOrderUtils.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }.toCharArray()
+
+    // /=Range(0x20, 0x7E)
+
+    fun average(left: String, right: String, alphabet: CharArray = DEFAULT_ALPHABET): String? {
+        return midPoints(left, right, 1, alphabet)?.firstOrNull()
+    }
+
+    fun midPoints(left: String, right: String, count: Int, alphabet: CharArray = DEFAULT_ALPHABET): List<String>? {
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/StringOrderUtils.kt:35]: Same remark about KDoc

File: vector/build.gradle [hunk 4]
```diff
@@ -365,6 +365,7 @@ dependencies {
     implementation "com.jakewharton.rxbinding3:rxbinding-appcompat:$rxbinding_version"
     implementation "com.jakewharton.rxbinding3:rxbinding-material:$rxbinding_version"
 
+//    implementation fileTree(include: ['*.jar', '*.aar'], dir: 'libs')
```
Comment (bmarty) [inline vector/build.gradle:368]: we do not have `libs` dir can you revert this change please?

File: vector/src/main/java/im/vector/app/features/spaces/SpaceSummaryController.kt [hunk 5]
```diff
@@ -139,7 +142,6 @@ class SpaceSummaryController @Inject constructor(
         }
 
         rootSpaces
-                ?.sortedBy { it.roomId }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/SpaceSummaryController.kt:142]: This line was crazy :)

**Review by bmarty — CHANGES_REQUESTED** [review]
LGTM, some small remark. Can you fix the issues reported by the CI please?
I will test on my device

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt (hunk 1 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt:249]: didn't know about the nullsLast, that's why I did the comparator

Comment (bmarty) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt:249]: `compareBy` looks very powerful, but a bit hard to read and understand what it actually does at first sight :)

Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt:249]: we have 5 or 6 Comparator that could be implemented with thenBy

Comment (Dominaezzz) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/SpaceOrderTest.kt]: This might be a tiny bit easier to read. (Comes with a hint of overhead but it's up to you)
```kotlin
compareBy<Pair<String, String?>> {""}
    .thenBy(nullsLast()) { it.second }
    .thenBy { it.first }
```

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/StringOrderUtils.kt (hunk 3 above)
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/util/StringOrderUtils.kt:35]: done

File: vector/build.gradle (hunk 4 above)
Comment (BillCarsonFr) [reply vector/build.gradle]: Done

**Review by bmarty — APPROVED** [review]
Thanks for the update, I merge it!
